source/crud/crud.rst

```diff
@@ -407,6 +407,17 @@ Read
 
   class FindOptions {
 
+    /**
+     * Enables writing to temporary files on the server. When set to true, the server
+     * can write temporary data to disk while executing the find operation.
+     *
+     * This option is sent only if the caller explicitly provides a value. The default
+     * is to not send a value.
```

Comment:
I didn't consider this in my original review weeks ago, but in light of [SPEC-1542](https://jira.mongodb.org/browse/SPEC-1542) (see: [this change](https://github.com/mongodb/specifications/commit/43f595ab4bb1594629d2a4f7a9041777781d8b64#diff-5efd2e4ce61f4ccda0ce031099303996R1818)) I think we should specifying what happens with this option in an unsupported server version. I think we have two options:

 * Add "For servers < 3.6, the driver MUST raise an error if the caller explicitly provides a value." This is because 3.6+ servers report errors for unknown command options. Raising a client-side error covers us for earlier server versions using the `find` command as well as pre-3.2 servers using OP_QUERY that have no way to express this option. 
 * Add a note that this option will be ignored for server versions < 3.6. This may entail adapting some text from the `maxAwaitTimeMS` doc block: "For servers < 3.2, this option is ignored and not sent as allowDiskUse does not exist in the OP_QUERY wire protocol." However, we'd still need to clarify that server versions >=3.2 and <3.6 will silently ignore the command option.

@jyemin might additionally request a test case for the client-side error reporting, as in https://github.com/mongodb/specifications/pull/650#pullrequestreview-332539817. If so, that could be a new test file similar to `find-allowdiskuse.yml` but with `error: true` included in each test case's `operation` block. Support for `allowDiskUse` was added in 4.3.1 by [SERVER-42077](https://jira.mongodb.org/browse/SERVER-42077), so we could either use "4.3.0" or something in the 3.4-range (e.g. "3.4.99") depending on whether we want to assert any error (server- and client-side) or just client-side (before 3.6 introduced strict checks for unknown options).

Reply:
allowDiskUse is similar to hint, isn't it?  Neither should change the result of the operation, but both can potentially have a large affect on the execution.  For allowDiskUse, it can mean that some executions will fail due to hitting the limit for in-memory sorts. For hint, it can mean that some executions may fail due to, say, a maxTimeExceeded error caused by the use of the wrong index.

So let's be consistent with hint and allowDiskUse.

Reply:
If I understand you correctly treating allowDiskUse consistently with how we treat hint means no further changes are needed. We only send allowDiskUse if the application specifies a value and we don't check server version.

Reply:
Done?

Reply:
@jmikola may need to correct me, but as I understand it, it means that a driver MUST throw a client side error if allowDiskUse is specified and server version is less than ... 4.4? or less than whatever server version started errors on unrecognized options.  At minimum it would mean throwing if server version is less than 3.2, since the old wire protocol certainly doesn't allow even trying to send allowDiskUse.

Reply:
That's the exact opposite of what I thought you were saying.

I think AllowDiskUse can (and should) be safely ignored if a server doesn't support it.

Its presence or absence NEVER affects the correctness of the results. In other words, you're going to get the exact same result set whether you specify AllowDiskUse or not (assuming sufficient memory).

Reply:
Here's the full matrix of scenarios as I see them:

1. Application does not set AllowDiskUse: nothing special to consider
2. Application sets AllowDiskUse and server ignores it: not a problem
3. Application sets AllowDiskUse and server returns an error because it doesn't recognize it: not a problem
4. Application sets AllowDiskUse and server ignores it: not a problem
5. Application sets AllowDiskUse and server supports it: intended behavior

Reply:
> allowDiskUse is similar to hint, isn't it? Neither should change the result of the operation if the operation succeeds, but both can potentially have a large affect on the execution.

@jyemin: I would not expect `allowDiskUse` to ever change the result of an operation, but I believe a `hint` may if it was used in conjunction with a sparse index since that could limit the pool of potential documents to match the query (an edge case I'll admit). 

----

> it means that a driver MUST throw a client side error if allowDiskUse is specified and server version is less than ... 4.4? or less than whatever server version started errors on unrecognized options. At minimum it would mean throwing if server version is less than 3.2, since the old wire protocol certainly doesn't allow even trying to send allowDiskUse.

I did some quick testing and it looks like server 3.2.21 raises an error for unrecognized options to the `find` command.

```
> db.runCommand({find:'foo',what:1})
{
	"waitedMS" : NumberLong(0),
	"ok" : 0,
	"errmsg" : "Failed to parse: { find: \"foo\", what: 1.0 }. Unrecognized field 'what'.",
	"code" : 9
}
```

Given that, I think we'd only need to throw client-side errors if the option would be used for a server version < 3.2 since that entails using OP_QUERY in the legacy manner without sending a `find` command.

Unfortunately, the server is inconsistent when it comes to reporting errors for unrecognized command options. It appears that validation for `find` was implemented when the command was introduced in 3.2, but as I discovered in [SPEC-1334](https://jira.mongodb.org/browse/SPEC-1334) validation for `update` didn't come until 3.4. Validation for `findAndModify` options was only introduced in 4.1.10 (see https://github.com/mongodb/specifications/pull/650#discussion_r328676881).

----

> Here's the full matrix of scenarios as I see them: ...

@rstam: Your list of scenarios above includes "Application sets AllowDiskUse and server ignores it: not a problem" twice. Was that a typo?

I agree that there's no need to test for the server returning an error if the option is unsupported (i.e. server versions >= [3.1.1](https://jira.mongodb.org/browse/SERVER-17282) and < [4.3.1](https://jira.mongodb.org/browse/SERVER-42077)). 

Per my original comment, I think raising an error is most consistent with other options we recently introduced, but I'm still OK with allowing it to be silently ignored if there's group preference for that. If we make a test, I'd propose using 3.1.0 as the `maxServerVersion`, since we know that's the last version of the server before the `find` command was introduced (alternatively, some suitably large 3.0.x version like 3.0.99 would suffice).

Reply:
> I'm still OK with allowing it to be silently ignored if there's group preference for that

To better clarify my last comment, if @jyemin agrees that `allowDiskUse` doesn't affect the result of a query (unlike `hint`, `collation`, and others) and is happy to ignore it when unsupported, I'll go along with that.

Going back to my first comment in this thread, "silently ignoring" is what we do for `maxAwaitTimeMS` before the `getMore` command, so one could argue this case is not much different -- both options affect _how_ the queries execute but not what documents they match.

Reply:
Seems like we're trading one error (client-side error that allowDiskUse not supported) for another that may happen only some times (hit 100 MB in-memory sort limit).  Especially as 3.0 is long since EOLed I'm fine with this outcome.

Reply:
@rstam: Since it seems we've settled on not enforcing any client-side error, I would propose we added the following language (akin to what we have for `maxAwaitTimeMS`):

> For servers < 3.2, this option is ignored and not sent as allowDiskUse does not exist in the OP_QUERY wire protocol.

I don't think we need to mention that this is only supported in MongoDB 4.4+, as the doc block for `collation` doesn't talk about when support was introduced (only when we want to raise client-side errors).